eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/PostRestartHtlcCleaner.scala

```diff
@@ -303,19 +308,71 @@ object PostRestartHtlcCleaner {
+                confirmedTxs.flatMap(tx => Closing.timedoutHtlcs(c.remoteCommit, d.commitments.remoteParams.dustLimit, tx, c.remoteCommitPublished))
+              case Some(c: Closing.NextRemoteClose) =>
+                val confirmedTxs = c.remoteCommitPublished.commitTx :: c.remoteCommitPublished.claimHtlcTimeoutTxs.filter(tx => irrevocablySpent.contains(tx.txid))
+                confirmedTxs.flatMap(tx => Closing.timedoutHtlcs(c.remoteCommit, d.commitments.remoteParams.dustLimit, tx, c.remoteCommitPublished))
+              case _ => Seq.empty[UpdateAddHtlc]
+            }).map(_.id).toSet
+            overriddenHtlcs ++ timedoutHtlcs
+          case _ => Set.empty
+        }
+        c.commitments.originChannels.filterKeys(htlcId => !htlcsToIgnore.contains(htlcId)).map { case (outgoingHtlcId, origin) => (origin, c.channelId, outgoingHtlcId) }
+      }
       .groupBy { case (origin, _, _) => origin }
-      .mapValues(_.map { case (_, channelId, htlcId) => (channelId, htlcId) })
+      .mapValues(_.map { case (_, channelId, htlcId) => (channelId, htlcId) }.toSet)
+      // We are only interested in HTLCs that are pending upstream (not fulfilled nor failed yet).
+      // It may be the case that we have unresolved HTLCs downstream that have been resolved upstream when the downstream
+      // channel is closing (e.g. due to an HTLC timeout) because cooperatively failing the HTLC downstream will be
+      // instant whereas the uncooperative close of the downstream channel will take time.
+      .filterKeys {
+        case _: Origin.Local => true
+        case Origin.Relayed(channelId, htlcId, _, _) => isPendingUpstream(channelId, htlcId)
+        case Origin.TrampolineRelayed(htlcs, _) => htlcs.exists { case (channelId, htlcId) => isPendingUpstream(channelId, htlcId) }
+      }
 
     val notRelayed = htlcsIn.filterNot(htlcIn => relayedOut.keys.exists(origin => matchesOrigin(htlcIn.add, origin)))
     log.info(s"htlcsIn=${htlcsIn.length} notRelayed=${notRelayed.length} relayedOut=${relayedOut.values.flatten.size}")
-    log.debug("notRelayed={}", notRelayed.map(htlc => (htlc.add.channelId, htlc.add.id)))
-    log.debug("relayedOut={}", relayedOut)
+    log.info("notRelayed={}", notRelayed.map(htlc => (htlc.add.channelId, htlc.add.id)))
+    log.info("relayedOut={}", relayedOut)
     BrokenHtlcs(notRelayed, relayedOut, Set.empty)
   }
 
+  /**
+   * List local channels that may have pending HTLCs, ignoring channels that are still in CLOSING state but have
+   * actually been closed. This can happen when the app is stopped just after a channel state has transitioned to CLOSED
+   * and before it has effectively been removed. Such closed channels will automatically be removed once the channel is
+   * restored.
+   */
+  private def listLocalChannels(channelsDb: ChannelsDb): Seq[HasCommitments] =
+    channelsDb.listLocalChannels().filter(c => Closing.isClosed(c, None).isEmpty)
```

Comment:
Nit, but in this particular instance I think the negation makes more sense

```suggestion
    channelsDb.listLocalChannels().filterNot(c => Closing.isClosed(c, None).isDefined)
```